File: Assignment 2: Syntax, Semantics, and Memory Management/Part1.2_Analyzing_Key_Features/multiplier.js

// JavaScript: Demonstrating Weak Typing and Scopes
function makeMultiplier(factor) {
    // Closure: Inner function retains access to 'factor'
    return function (number) {
        return number * factor;
    };
}

const double = makeMultiplier(2);
const triple = makeMultiplier(3);

console.log(double(5)); // Outputs 10
console.log(triple(5)); // Outputs 15

// Weak typing example: Implicit type conversion
let value = "10" - 2; // Coerces "10" to number
console.log(value); // Outputs 8

let anotherValue = "10" + 2; // Coerces 2 to string
console.log(anotherValue); // Outputs "102"

File: Assignment 2: Syntax, Semantics, and Memory Management/Part1.2_Analyzing_Key_Features/multiplier.py
def make_multiplier(factor):
    def multiplier(number):
        return number * factor
    return multiplier

double = make_multiplier(2)
triple = make_multiplier(3)

print(double(5))  # Outputs 10
print(triple(5))  # Outputs 15 
# Dynamic typing: Reassign variable to a different type
double = "I am now a string"
print(double)
